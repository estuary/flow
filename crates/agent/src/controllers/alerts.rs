use std::collections::HashMap;

use chrono::{DateTime, Utc};
use models::status::{AlertState, AlertType, Alerts, ControllerAlert};

pub fn set_alert_firing(
    statuses: &mut Alerts,
    alert_type: AlertType,
    now: DateTime<Utc>,
    error: String,
    count: u32,
    spec_type: models::CatalogType,
) {
    // If the current status is for a resolved alert of the same type, remove
    // that entry and re-start with a fresh alert.
    if statuses
        .get(&alert_type)
        .is_some_and(|s| s.state == AlertState::Resolved)
    {
        statuses.remove(&alert_type);
    }

    if let Some(existing) = statuses.get_mut(&alert_type) {
        existing.last_ts = Some(now);
        existing.count = count;
        existing.spec_type = spec_type;
        existing.error = error;
    } else {
        tracing::info!(%alert_type, %spec_type, "alert started firing");
        statuses.insert(
            alert_type,
            ControllerAlert {
                state: AlertState::Firing,
                first_ts: now,
                last_ts: None,
                error,
                count,
                spec_type,
                resolved_at: None,
                extra: HashMap::new(),
            },
        );
    }
}

/// Clears the given alert entirely, removing it from `statuses`.
/// This does not allow for the presence of separate `resolved_arguments`.
/// If separate `resovlved_arguments` are needed, then we'll have to
pub fn resolve_alert(statuses: &mut Alerts, alert_type: AlertType) {
    if let Some(_cleared) = statuses.remove(&alert_type) {
        tracing::info!(%alert_type, "alert resolved");
    }
}
