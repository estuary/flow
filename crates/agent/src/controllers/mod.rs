mod capture;
mod catalog_test;
mod collection;
mod handler;
mod materialization;
mod publication_status;

use crate::controlplane::ControlPlane;
use anyhow::Context;
use chrono::{DateTime, Utc};
use models::{AnySpec, CatalogType, Id};
use proto_flow::{flow, AnyBuiltSpec};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use sqlx::types::Uuid;
use std::fmt::Debug;

use self::{
    capture::CaptureStatus, catalog_test::TestStatus, collection::CollectionStatus,
    materialization::MaterializationStatus,
};

pub use handler::ControllerHandler;

/// This version is used to determine if the controller state is compatible with the current
/// code. Any controller state having a higher version than this will be ignored.
pub const CONTROLLER_VERSION: i32 = 2;

/// Represents the state of a specific controller and catalog_name.
#[derive(Clone, Debug, Serialize)]
pub struct ControllerState {
    pub catalog_name: String,
    /// The live spec corresponding to this controller, which will be `None` if
    /// the spec has been deleted.
    pub live_spec: Option<AnySpec>,
    /// The built spec that goes along with the live spec, which will be `None`
    /// if the spec has been deleted.
    pub built_spec: Option<AnyBuiltSpec>,
    /// The current `controller_next_run` value. This is useful for knowing
    /// when the controller run was desired, which may have been earlier than
    /// the actual time of the current run.
    pub next_run: Option<DateTime<Utc>>,
    /// The last update time of the controller.
    pub updated_at: DateTime<Utc>,
    /// The number of consecutive failures from previous controller runs. This
    /// gets reset to 0 after any successful controller run.
    pub failures: i32,
    /// The error output from the most recent controller run, or `None` if the
    /// most recent run was successful. If `error` is `Some`, then `failures`
    /// will be > 0.
    pub error: Option<String>,
    /// The `last_pub_id` of the corresponding `live_specs` row.
    pub last_pub_id: Id,
    /// The `last_build_id` of the corresponding `live_specs` row.
    pub last_build_id: Id,
    /// The logs token that's used for all operations of this controller. Every
    /// run of a given controller uses the same `logs_token` so that you can
    /// see all the logs in one place.
    pub logs_token: Uuid,
    /// The version of this controller's `current_status`, which will always be
    /// less than or equal to `CONTROLLER_VERSION`.
    pub controller_version: i32,
    /// The current `status` of the controller, which represents the before
    /// state during an update. This is just informational.
    pub current_status: Status,
    /// ID of the data plane in which this specification lives.
    pub data_plane_id: Id,
    /// The `dependency_hash` of the `live_specs` row, used to determine whether any
    /// dependencies have had their models changed.
    pub live_dependency_hash: Option<String>,
}

impl ControllerState {
    pub fn parse_db_row(
        job: &agent_sql::controllers::ControllerJob,
    ) -> anyhow::Result<ControllerState> {
        let status: Status = if job.controller_version == 0 {
            Status::Uninitialized
        } else {
            serde_json::from_str(job.status.get()).context("deserializing controller status")?
        };

        // Spec_type may be null for specs last published by a previous version.
        // We now leave the spec_type in place when soft-deleting live_specs.
        let maybe_type = job.spec_type.map(Into::<CatalogType>::into);
        let (live_spec, built_spec) = if let Some(catalog_type) = maybe_type {
            let live_spec = if let Some(live_json) = &job.live_spec {
                let spec = AnySpec::deserialize(catalog_type, live_json.get())
                    .context("deserializing live spec")?;
                Some(spec)
            } else {
                None
            };

            let built_spec = if let Some(built_json) = &job.built_spec {
                let s = match catalog_type {
                    CatalogType::Capture => {
                        AnyBuiltSpec::Capture(serde_json::from_str::<flow::CaptureSpec>(
                            built_json.get(),
                        )?)
                    }
                    CatalogType::Collection => {
                        AnyBuiltSpec::Collection(serde_json::from_str::<flow::CollectionSpec>(
                            built_json.get(),
                        )?)
                    }
                    CatalogType::Materialization => {
                        AnyBuiltSpec::Materialization(serde_json::from_str::<
                            flow::MaterializationSpec,
                        >(built_json.get())?)
                    }
                    CatalogType::Test => AnyBuiltSpec::Test(
                        serde_json::from_str::<flow::TestSpec>(built_json.get())?,
                    ),
                };
                Some(s)
            } else {
                None
            };
            if live_spec.is_some() != built_spec.is_some() {
                anyhow::bail!(
                    "expected live and built specs to both be Some or None, got live: {}, built: {}",
                    live_spec.is_some(),
                    built_spec.is_some()
                );
            }

            (live_spec, built_spec)
        } else {
            (None, None)
        };

        let controller_state = ControllerState {
            next_run: job.controller_next_run,
            updated_at: job.updated_at,
            live_spec,
            built_spec,
            failures: job.failures,
            catalog_name: job.catalog_name.clone(),
            error: job.error.clone(),
            last_pub_id: job.last_pub_id.into(),
            last_build_id: job.last_build_id.into(),
            logs_token: job.logs_token,
            controller_version: job.controller_version,
            current_status: status,
            data_plane_id: job.data_plane_id.into(),
            live_dependency_hash: job.live_dependency_hash.clone(),
        };
        Ok(controller_state)
    }
}

/// A wrapper around an `anyhow::Error` that also contains retry information.
/// Allows marking an error as non-retryable by setting `retry` to `None`.
#[derive(Debug)]
pub struct RetryableError {
    pub inner: anyhow::Error,
    pub retry: Option<NextRun>,
}

impl std::fmt::Display for RetryableError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let maybe_retry = if self.retry.is_some() {
            "(will retry)"
        } else {
            "(terminal error)"
        };
        write!(f, "{} {}", self.inner, maybe_retry)
    }
}

impl std::error::Error for RetryableError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        self.inner.source()
    }
}

/// A trait to help controllers specify retry behavior for different errors.
/// Allows calling `my_result.with_retry(next_run)?` in order to consider an
/// error retriable.
trait ControllerErrorExt {
    type Success;

    fn with_maybe_retry(
        self,
        maybe_retry: Option<NextRun>,
    ) -> Result<Self::Success, RetryableError>;

    fn with_retry(self, after: NextRun) -> Result<Self::Success, RetryableError>
    where
        Self: Sized,
    {
        self.with_maybe_retry(Some(after))
    }

    fn do_not_retry(self) -> Result<Self::Success, RetryableError>
    where
        Self: Sized,
    {
        self.with_maybe_retry(None)
    }
}

impl<T, E: Into<anyhow::Error>> ControllerErrorExt for Result<T, E> {
    type Success = T;
    fn with_maybe_retry(self, after: Option<NextRun>) -> Result<T, RetryableError> {
        self.map_err(|e| RetryableError {
            inner: e.into(),
            retry: after,
        })
    }
}

/// Represents a desired future run of the controller.
/// This is represented as a simple duration and jitter in order to make
/// testing easier, and to keep controller implementations simple.
#[derive(Debug, Serialize, Clone, Copy, PartialEq, Eq)]
pub struct NextRun {
    pub after_seconds: u32,
    pub jitter_percent: u16,
}

impl std::cmp::PartialOrd for NextRun {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl std::cmp::Ord for NextRun {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.after_seconds
            .cmp(&other.after_seconds)
            .then(self.jitter_percent.cmp(&other.jitter_percent))
    }
}

impl NextRun {
    const DEFAULT_JITTER: u16 = 20;

    pub fn after_minutes(minutes: u32) -> NextRun {
        NextRun {
            after_seconds: minutes * 60,
            jitter_percent: NextRun::DEFAULT_JITTER,
        }
    }

    pub fn with_jitter_percent(self, jitter_percent: u16) -> Self {
        NextRun {
            after_seconds: self.after_seconds,
            jitter_percent,
        }
    }

    /// Returns an absolute time at which the next run should become due.
    /// Uses only millisecond precision to ensure that the timestamp can be losslessly
    /// round-tripped through postgres.
    pub fn compute_time(&self) -> DateTime<Utc> {
        use rand::Rng;

        let delta_millis = self.after_seconds as i64 * 1000;
        let jitter_mul = self.jitter_percent as f64 / 100.0;
        let jitter_max = (delta_millis as f64 * jitter_mul) as i64;
        let jitter_add = rand::thread_rng().gen_range(0..jitter_max);
        let dur = chrono::TimeDelta::milliseconds(delta_millis + jitter_add);
        Utc::now() + dur
    }
}

/// Represents the internal state of a controller.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, JsonSchema)]
#[serde(tag = "type")]
pub enum Status {
    Capture(CaptureStatus),
    Collection(CollectionStatus),
    Materialization(MaterializationStatus),
    Test(TestStatus),
    #[schemars(skip)]
    #[serde(other, untagged)]
    Uninitialized,
}

/// Returns a backoff after failing to activate or delete shards/journals in the
/// data-plane. Failures to do so should be re-tried indefinitely.
fn backoff_data_plane_activate(prev_failures: i32) -> NextRun {
    let after_minutes = if prev_failures < 3 {
        prev_failures.max(1) as u32
    } else {
        prev_failures as u32 * 15
    };
    NextRun::after_minutes(after_minutes)
}

impl Status {
    pub fn json_schema() -> schemars::schema::RootSchema {
        let settings = schemars::gen::SchemaSettings::draft2019_09();
        //settings.option_add_null_type = false;
        //settings.inline_subschemas = true;
        let generator = schemars::gen::SchemaGenerator::new(settings);
        generator.into_root_schema_for::<Status>()
    }

    fn catalog_type(&self) -> Option<CatalogType> {
        match self {
            Status::Capture(_) => Some(CatalogType::Capture),
            Status::Collection(_) => Some(CatalogType::Collection),
            Status::Materialization(_) => Some(CatalogType::Materialization),
            Status::Test(_) => Some(CatalogType::Test),
            Status::Uninitialized => None,
        }
    }

    /// The main logic of a controller run is performed as an update of the status.
    async fn update<C: ControlPlane>(
        &mut self,
        state: &ControllerState,
        control_plane: &mut C,
    ) -> anyhow::Result<Option<NextRun>> {
        let Some(live_spec) = &state.live_spec else {
            // There's no need to delete tests and nothing depends on them.
            if let Some(catalog_type) = self.catalog_type().filter(|ct| *ct != CatalogType::Test) {
                // The live spec has been deleted. Delete the data plane
                // resources, and then notify dependent controllers, to make
                // sure that they can respond. The controller job row will be
                // deleted automatically after we return.
                control_plane
                    .data_plane_delete(
                        state.catalog_name.clone(),
                        catalog_type,
                        state.data_plane_id,
                    )
                    .await
                    .context("deleting from data-plane")
                    .with_retry(backoff_data_plane_activate(state.failures))?;
                tracing::info!("deleted from data-plane");
                control_plane
                    .notify_dependents(state.catalog_name.clone())
                    .await
                    .expect("failed to update dependents");
            } else {
                tracing::info!("skipping data-plane deletion because there is no spec_type");
            }
            return Ok(None);
        };

        let next_run = match live_spec {
            AnySpec::Capture(c) => {
                let capture_status = self.as_capture_mut()?;
                capture_status.update(state, control_plane, c).await?
            }
            AnySpec::Collection(c) => {
                let collection_status = self.as_collection_mut()?;
                collection_status.update(state, control_plane, c).await?
            }
            AnySpec::Materialization(m) => {
                let materialization_status = self.as_materialization_mut()?;

                materialization_status
                    .update(state, control_plane, m)
                    .await?
            }
            AnySpec::Test(t) => {
                let test_status = self.as_test_mut()?;
                test_status.update(state, control_plane, t).await?
            }
        };
        tracing::info!(?next_run, "finished controller update");
        Ok(next_run)
    }

    pub fn is_uninitialized(&self) -> bool {
        matches!(self, Status::Uninitialized)
    }

    fn as_capture_mut(&mut self) -> anyhow::Result<&mut CaptureStatus> {
        if self.is_uninitialized() {
            *self = Status::Capture(Default::default());
        }
        match self {
            Status::Capture(c) => Ok(c),
            _ => anyhow::bail!("expected capture status"),
        }
    }

    fn as_collection_mut(&mut self) -> anyhow::Result<&mut CollectionStatus> {
        if self.is_uninitialized() {
            *self = Status::Collection(Default::default());
        }
        match self {
            Status::Collection(c) => Ok(c),
            _ => anyhow::bail!("expected collection status"),
        }
    }

    fn as_materialization_mut(&mut self) -> anyhow::Result<&mut MaterializationStatus> {
        if self.is_uninitialized() {
            *self = Status::Materialization(Default::default());
        }
        match self {
            Status::Materialization(m) => Ok(m),
            _ => anyhow::bail!("expected materialization status"),
        }
    }

    fn as_test_mut(&mut self) -> anyhow::Result<&mut TestStatus> {
        if self.is_uninitialized() {
            *self = Status::Test(Default::default());
        }
        match self {
            Status::Test(t) => Ok(t),
            _ => anyhow::bail!("expected test status"),
        }
    }

    #[cfg(test)]
    pub fn unwrap_capture(&self) -> &CaptureStatus {
        match self {
            Status::Capture(c) => c,
            _ => panic!("expected capture status"),
        }
    }

    #[cfg(test)]
    pub fn unwrap_collection(&self) -> &CollectionStatus {
        match self {
            Status::Collection(c) => c,
            _ => panic!("expected collection status"),
        }
    }

    #[cfg(test)]
    pub fn unwrap_materialization(&self) -> &MaterializationStatus {
        match self {
            Status::Materialization(m) => m,
            _ => panic!("expected materialization status"),
        }
    }

    #[cfg(test)]
    pub fn unwrap_test(&self) -> &TestStatus {
        match self {
            Status::Test(t) => t,
            _ => panic!("expected test status"),
        }
    }
}

fn datetime_schema(_: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
    serde_json::from_value(serde_json::json!({
        "type": "string",
        "format": "date-time",
    }))
    .unwrap()
}

#[cfg(test)]
mod test {
    use std::collections::{BTreeSet, VecDeque};

    use chrono::TimeZone;

    use super::*;
    use crate::controllers::materialization::SourceCaptureStatus;
    use crate::controllers::publication_status::{
        ActivationStatus, PublicationInfo, PublicationStatus,
    };
    use crate::draft::Error;
    use crate::publications::{AffectedConsumer, IncompatibleCollection, JobStatus, RejectedField};

    #[test]
    fn test_status_round_trip_serde() {
        let mut add_bindings = BTreeSet::new();
        add_bindings.insert(models::Collection::new("snails/shells"));

        let pub_status = PublicationInfo {
            id: Id::new([4, 3, 2, 1, 1, 2, 3, 4]),
            created: Some(Utc.with_ymd_and_hms(2024, 5, 30, 9, 10, 11).unwrap()),
            completed: Some(Utc.with_ymd_and_hms(2024, 5, 30, 9, 10, 11).unwrap()),
            detail: Some("some detail".to_string()),
            result: Some(JobStatus::build_failed(vec![IncompatibleCollection {
                collection: "snails/water".to_string(),
                requires_recreation: Vec::new(),
                affected_materializations: vec![AffectedConsumer {
                    name: "snails/materialize".to_string(),
                    fields: vec![RejectedField {
                        field: "a_field".to_string(),
                        reason: "do not like".to_string(),
                    }],
                }],
            }])),
            errors: vec![Error {
                catalog_name: "snails/shells".to_string(),
                scope: Some("flow://materializations/snails/shells".to_string()),
                detail: "a_field simply cannot be tolerated".to_string(),
            }],
            count: 1,
            is_touch: false,
        };
        let mut history = VecDeque::new();
        history.push_front(pub_status);

        let status = Status::Materialization(MaterializationStatus {
            activation: ActivationStatus {
                last_activated: Id::new([1, 2, 3, 4, 4, 3, 2, 1]),
            },
            source_capture: Some(SourceCaptureStatus {
                up_to_date: false,
                add_bindings,
            }),
            publications: PublicationStatus {
                max_observed_pub_id: Id::new([1, 2, 3, 4, 5, 6, 7, 8]),
                history,
                dependency_hash: Some("abc12345".to_string()),
            },
        });

        let as_json = serde_json::to_string_pretty(&status).expect("failed to serialize status");
        let round_tripped: Status =
            serde_json::from_str(&as_json).expect("failed to deserialize status");

        #[derive(Debug)]
        #[allow(unused)]
        struct StatusSnapshot {
            starting: Status,
            json: String,
            parsed: Status,
        }

        insta::assert_debug_snapshot!(
            "materialization-status-round-trip",
            StatusSnapshot {
                starting: status,
                json: as_json,
                parsed: round_tripped,
            }
        );
    }

    #[test]
    fn test_status_json_schema() {
        let schema = serde_json::to_value(Status::json_schema()).unwrap();
        insta::assert_json_snapshot!(schema);
    }
}
