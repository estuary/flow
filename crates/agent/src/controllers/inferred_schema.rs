use std::collections::{BTreeMap, BTreeSet};

use crate::controllers::publication_status::{PublicationHistory, PublicationStatus};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use tables::{DraftRow, EitherOrBoth, SpecExt};

use super::{
    ControlJob, ControlPlane, ControllerState, ControllerUpdate, NextRun, PublicationResult,
};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct InferredSchemaStatus {
    pub schema_last_updated: DateTime<Utc>,
    pub schema_md5: Option<String>,
    pub publications: PublicationHistory,
}

impl InferredSchemaStatus {
    fn next_run(&self, now: DateTime<Utc>) -> NextRun {
        // The idea here is to check frequently if there isn't an inferred schema at all yet,
        // so we can quickly start materializing some data. But after it works at least once,
        // then we want to use a longer duration in order to coalesce more schema updates into
        // each publication to prevent undue churn.
        // TODO: we might want to account for `last_backfill` times here
        let min_backoff_minutes = if self.schema_md5.is_none() {
            1i64
        } else {
            10i64
        };

        let after_minutes = now
            .signed_duration_since(self.schema_last_updated)
            .num_minutes()
            .max(min_backoff_minutes)
            .min(90);
        NextRun::after_minutes(after_minutes as u32).with_jitter_percent(25)
    }

    fn is_publication_pending(&self) -> bool {
        self.publications.pending.is_some()
    }

    /// Updates the status of all consumers of the collection to reflect that they're using the latest inferred schema.
    /// This works because if the collection itself is published, then all consumers of it must also be included in the
    /// publication.
    fn on_successful_collection_publication(
        &self,
        _collection_name: &models::Collection,
        schema_md5: Option<String>,
        publication: &PublicationResult,
    ) -> Self {
        let mut new_status = self.clone();
        new_status.schema_md5 = schema_md5.clone();
        new_status.schema_last_updated = publication.completed_at;

        let pub_status = PublicationStatus::observed(publication);
        let _completed = new_status.publications.observe(pub_status.clone());

        new_status
    }

    fn on_failed_publication(
        &self,
        collection_name: &models::Collection,
        publication: &PublicationResult,
    ) -> Self {
        let mut new_status = self.clone();
        let schema_md5 = get_inferred_schema_md5(collection_name, publication);
        // We can still update the last known schema md5, if known.
        if schema_md5.is_some() && schema_md5 != new_status.schema_md5 {
            new_status.schema_md5 = schema_md5.clone();
            new_status.schema_last_updated = publication.completed_at;
        }

        let _ = new_status
            .publications
            .observe(PublicationStatus::observed(publication))
            .expect("publication must be pending");

        new_status
    }
}

fn get_inferred_schema_md5(
    collection: &models::Collection,
    publication: &PublicationResult,
) -> Option<String> {
    publication
        .live
        .inferred_schemas
        .get_by_key(collection)
        .map(|s| s.md5.clone())
}

pub struct InferredSchemaController;

#[async_trait::async_trait]
impl ControlJob for InferredSchemaController {
    type Status = InferredSchemaStatus;

    fn controller_name(&self) -> String {
        String::from("InferredSchemaPublisher")
    }

    fn observe_publication(
        &self,
        current_states: BTreeMap<String, ControllerState<Self::Status>>,
        publication: &PublicationResult,
    ) -> anyhow::Result<BTreeMap<String, ControllerUpdate<Self::Status>>> {
        if publication.publication_status.is_success() {
            Ok(on_successful_publication(&current_states, publication))
        } else {
            Ok(on_failed_publication(&current_states, publication))
        }
    }

    async fn update(
        &self,
        catalog_name: String,
        current_state: ControllerState<Self::Status>,
        control_plane: &mut dyn ControlPlane,
    ) -> anyhow::Result<ControllerUpdate<Self::Status>> {
        if current_state.status.is_publication_pending() {
            // We're still waiting on a publication to complete, so just keep waiting.
            // Note that we don't set next_run or update the status.
            return Ok(current_state.to_update());
        }
        let collection_name = models::Collection::new(catalog_name);

        let maybe_schema_md5 = dbg!(control_plane
            .get_inferred_schema(collection_name.clone())
            .await?
            .map(|s| s.md5));

        if current_state.status.schema_md5 != maybe_schema_md5 {
            let mut new_status = current_state.status.clone();
            create_publication(collection_name, &mut new_status, control_plane).await?;

            Ok(current_state.to_update().with_status(new_status))
        } else {
            eprintln!("wtf everything is up to date");
            // Everything seems up to date, so just schedule the next check-up
            let next_run = current_state.status.next_run(control_plane.current_time());
            Ok(current_state.to_update().with_next_run(next_run))
        }
    }
}

async fn create_publication(
    collection_name: models::Collection,
    new_status: &mut InferredSchemaStatus,
    control_plane: &mut dyn ControlPlane,
) -> anyhow::Result<()> {
    let mut names = BTreeSet::new();
    names.insert(collection_name.clone());
    let collections = control_plane.get_collections(names).await?;
    if collections.is_empty() {
        anyhow::bail!("collection '{collection_name}' not found");
    };

    let catalog = tables::LiveCatalog {
        collections,
        ..Default::default()
    };
    let pub_id = control_plane
        .create_publication(tables::DraftCatalog::from(catalog))
        .await?;
    new_status.publications.pending = Some(PublicationStatus::created(
        pub_id,
        control_plane.current_time(),
    ));
    Ok(())
}

fn on_successful_publication(
    current_states: &BTreeMap<String, ControllerState<InferredSchemaStatus>>,
    publication: &PublicationResult,
) -> BTreeMap<String, ControllerUpdate<InferredSchemaStatus>> {
    let mut updates = BTreeMap::new();

    let drafted_collections = publication
        .draft
        .collections
        .iter()
        .filter(|c| c.spec.as_ref().map(|s| s.is_enabled()).unwrap_or(false))
        .map(|c| (c.catalog_name(), c.spec().unwrap()));

    fn left_outer<L, R>(eob: EitherOrBoth<L, R>) -> Option<(L, Option<R>)> {
        match eob {
            EitherOrBoth::Both(l, r) => Some((l, Some(r))),
            EitherOrBoth::Left(l) => Some((l, None)),
            EitherOrBoth::Right(_) => None,
        }
    }

    fn right_outer<L, R>(eob: EitherOrBoth<L, R>) -> Option<(Option<L>, R)> {
        match eob {
            EitherOrBoth::Both(l, r) => Some((Some(l), r)),
            EitherOrBoth::Left(l) => None,
            EitherOrBoth::Right(r) => Some((None, r)),
        }
    }

    for (maybe_inferred_schema, (collection_name, draft_spec)) in publication
        .live
        .inferred_schemas
        .outer_join(drafted_collections, right_outer)
    {
        let current_state = current_states.get(collection_name.as_str());
        let desired_active = uses_inferred_schema(draft_spec);
        let schema_md5 = maybe_inferred_schema.as_ref().map(|s| s.md5.clone());

        let maybe_update = match (current_state, desired_active) {
            (Some(state), true) => {
                let next_status = state.status.on_successful_collection_publication(
                    collection_name,
                    schema_md5,
                    publication,
                );
                let next_run = next_status.next_run(publication.completed_at);
                Some(
                    state
                        .to_update()
                        .with_status(next_status)
                        .with_next_run(next_run),
                )
            }
            // State is currently active, but the collection no longer uses schema inference, so we should deactivate.
            (Some(state), false) if state.active => Some(state.to_update().with_active(false)),
            (Some(_), false) => None, // inactive state remains inactive
            // Need to initialize a new state for this collection
            (None, true) => {
                let next_status = InferredSchemaStatus {
                    schema_md5,
                    schema_last_updated: publication.completed_at,
                    publications: PublicationHistory::initial(PublicationStatus::observed(
                        publication,
                    )),
                };
                let next_run = next_status.next_run(publication.completed_at);
                Some(ControllerUpdate::new(next_status).with_next_run(next_run))
            }
            (None, false) => None,
        };

        if let Some(update) = maybe_update {
            updates.insert(collection_name.to_string(), update);
        }
    }

    updates
}

fn on_failed_publication(
    current_states: &BTreeMap<String, ControllerState<InferredSchemaStatus>>,
    publication: &PublicationResult,
) -> BTreeMap<String, ControllerUpdate<InferredSchemaStatus>> {
    let mut updates = BTreeMap::new();

    for (collection_name, state) in current_states.iter().filter(|(_, v)| v.active) {
        if state
            .status
            .publications
            .is_pending(publication.publication_id)
        {
            let collection = models::Collection::new(collection_name);
            let next_status = state.status.on_failed_publication(&collection, publication);
            let next_run = next_status.next_run(publication.completed_at);
            let update = state
                .to_update()
                .with_status(next_status)
                .with_next_run(next_run);
            updates.insert(collection_name.clone(), update);
        }
    }

    updates
}

fn uses_inferred_schema(collection: &models::CollectionDef) -> bool {
    collection
        .read_schema
        .as_ref()
        .map(models::Schema::references_inferred_schema)
        .unwrap_or(false)
}

#[cfg(test)]
mod test {
    use crate::controllers::test_util::{self, Harness, TestPublication};
    use serde_json::{json, Value};

    use super::*;

    #[test]
    fn happy_path_test() {
        let mut harness = Harness::new(InferredSchemaController);

        let initial_draft = test_util::draft_of(json!({
            "captures": {
                "a/cap": mock_capture_spec(&["acmeCo/a"]),
            },
            "collections": {
                "a/a": mock_collection_spec(true),
                "a/d": mock_derivation_spec(true, &["a/a"]),
                // f should be ignored because it doesn't use schema inference
                "a/f": mock_collection_spec(false),
            },
            "materializations": {
                "a/m1": mock_materialization_spec(&["a/a", "a/d"]),
                "a/m2": mock_materialization_spec(&["a/a"]),
                "a/m3": mock_materialization_spec(&["a/d"]),
            }
        }));

        let (info, updates) = harness.observe_publication(TestPublication::of_draft(initial_draft));
        // Expect to see updates for both "a/a" and "a/d" that initialize new controller states
        // for those collections.
        insta::with_settings!({ info => &info }, {
            insta::assert_json_snapshot!("initial-updates", &updates);
        });

        // The inferred schema of a/a gets updated.
        harness.update_inferred_schema("a/a", 1);

        // Technically, either a/a or a/d could be run next in the real world, since they both have
        // identical next_run values. Time is deterministic in the test environment, though, and the
        // harness always selects controllers in lexicographical order.
        let (info, update, mut publications) = harness.next_run_update();
        assert_eq!(info.catalog_name, "a/a");
        // Expect to see that a publication of a/a was created in response to the schema being updated
        insta::with_settings!({ info => &info }, {
            insta::assert_json_snapshot!("update-after-a-schema-updated-1", (update, &publications));
        });

        // Now the controller observes the successful completion of that publication
        let publication = publications.pop().unwrap();
        let (info, updates) = harness.observe_publication(publication);

        // Expect to see the successful publication in the history of a/a
        insta::with_settings!({ info => &info }, {
            insta::assert_json_snapshot!("observe-pub-a-1-completed", &updates);
        });

        // Now the inferred schema for a/d gets updated
        harness.update_inferred_schema("a/d", 1);

        let (info, update, mut publications) = harness.next_run_update();

        assert_eq!(info.catalog_name, "a/d");
        // Expect to see that a publication of a/d was created in response to the schema being updated
        insta::with_settings!({ info => &info }, {
            insta::assert_json_snapshot!("update-after-d-schema-updated-1", (update, &publications));
        });

        let (info, updates) = harness.observe_publication(publications.pop().unwrap());

        // Expect to see the successful publication in the history of a/d
        insta::with_settings!({ info => &info }, {
            insta::assert_json_snapshot!("observe-pub-d-1-completed", &updates);
        });

        // Assert that the next update runs don't publish anthing, since the inferred schemas haven't been updated
        let (info, update, publications) = harness.next_run_update();
        assert_eq!(info.catalog_name, "a/a");
        assert!(publications.is_empty());
        assert!(update.next_run.is_some());

        let (info, update, publications) = harness.next_run_update();
        assert_eq!(info.catalog_name, "a/d");
        assert!(publications.is_empty());
        assert!(update.next_run.is_some());
    }

    fn mock_derivation_spec(use_inferred: bool, sources: &[&str]) -> Value {
        let mut collection = mock_collection_spec(use_inferred);

        let transforms = sources
            .iter()
            .map(|name| {
                json!({
                    "name": name, "source": name, "lambda": "shelect shtar"
                })
            })
            .collect::<Vec<_>>();
        let derive = json!({
            "using": {
                "sqlite": {}
            },
            "transforms": transforms,
        });
        collection
            .as_object_mut()
            .unwrap()
            .insert("derive".to_owned(), derive);
        collection
    }

    fn mock_collection_spec(uses_inferred: bool) -> Value {
        let read_schema = if uses_inferred {
            json!({
                "allOf": [
                    { "$ref": "flow://inferred-schema" },
                    { "$ref": "flow://write-schema" },
                ]
            })
        } else {
            json!({"type": "object"})
        };
        json!({
            "key": ["/id"],
            "writeSchema": {
                "type": "object",
                "properties": {
                    "id": {"type": "string"}
                }
            },
            "readSchema": read_schema,
        })
    }

    fn mock_capture_spec(bound_collections: &[&str]) -> Value {
        let bindings = bound_collections
            .into_iter()
            .map(|collection| {
                json!({
                    "resource": { "thing": collection },
                    "target": collection,
                })
            })
            .collect();
        mock_connector_task(bindings)
    }

    fn mock_materialization_spec(bound_collections: &[&str]) -> Value {
        let bindings = bound_collections
            .into_iter()
            .map(|collection| {
                json!({
                    "resource": { "thing": collection },
                    "source": collection,
                })
            })
            .collect();
        mock_connector_task(bindings)
    }

    fn mock_connector_task(bindings: Vec<Value>) -> Value {
        serde_json::json!({
            "endpoint": {
                "connector": {
                    "image": "foo:test",
                    "config": {},
                }
            },
            "bindings": bindings,
        })
    }
}
