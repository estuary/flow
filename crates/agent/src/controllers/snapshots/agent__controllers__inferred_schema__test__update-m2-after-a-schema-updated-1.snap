---
source: crates/agent/src/controllers/inferred_schema.rs
expression: "(update, &publications)"
info:
  catalog_name: a/m2
  prev_state:
    active: true
    next_run: "2024-01-01T08:02:00Z"
    updated_at: "2024-01-01T08:01:00Z"
    status:
      collections:
        a/a:
          inferred_schema_md5: ~
          md5_last_updated: "2024-01-01T08:01:00Z"
      publications:
        pending: ~
        history: []
        last_observed:
          id: "0100000000000000"
          completed: "2024-01-01T08:01:00Z"
          result:
            type: success
    failures: 0
---
[
  {
    "active": true,
    "next_run": null,
    "status": {
      "collections": {
        "a/a": {
          "inferred_schema_md5": null,
          "md5_last_updated": "2024-01-01T08:01:00Z"
        }
      },
      "publications": {
        "pending": {
          "id": "0300000000000000",
          "created": "2024-01-01T08:02:00Z"
        },
        "history": [],
        "last_observed": {
          "id": "0100000000000000",
          "completed": "2024-01-01T08:01:00Z",
          "result": {
            "type": "success"
          }
        }
      }
    }
  },
  [
    {
      "draft": {
        "materializations": {
          "a/m2": {
            "endpoint": {
              "connector": {
                "image": "foo:test",
                "config": {
                  "$serde_json::private::RawValue": "{}"
                }
              }
            },
            "bindings": [
              {
                "resource": {
                  "$serde_json::private::RawValue": "{\"thing\":\"a/a\"}"
                },
                "source": "a/a",
                "fields": {
                  "recommended": true
                }
              }
            ]
          }
        }
      },
      "publication_meta": [
        "0300000000000000",
        "2024-01-01T08:02:00Z"
      ]
    }
  ]
]
